In [3]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn import preprocessing
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score, RepeatedKFold, KFold, StratifiedKFold
from sklearn.metrics import precision_score, recall_score, f1_score

In [6]:
data=pd.read_csv('Coimbra_breast_cancer_dataset.csv')
df=data.copy()
df

Age        BMI  Glucose  Insulin      HOMA   Leptin  Adiponectin  \
0      54  35.207389      103    5.642  1.378660  65.6699     9.738408   
1      52  22.978520      132    6.054  1.145435  47.5445     3.627241   
2      32  21.101341       87    5.668  1.008595  50.5074     5.067841   
3      42  26.761205      132    2.875  1.003837  16.8972    10.096475   
4      55  34.232520       76    3.120  1.597721  17.6852    11.845054   
...   ...        ...      ...      ...       ...      ...          ...   
3995   83  24.222881       76   10.511  0.733816  17.9836     5.423659   
3996   84  21.504832      137    4.567  1.314829  28.1327     5.878012   
3997   66  29.198300      137    6.823  0.688278   8.7539    10.242825   
3998   73  22.871102       90   10.564  3.623631  26.8907    14.407806   
3999   34  28.027362       76    3.008  1.235593  51.8521     7.984090   

      Resistin    MCP.1  Classification  
0     31.17499  197.660               1  
1     23.03327  423.361               2  
2      9.51156  897.034               2  
3      9.75652  267.744               1  
4     17.21541  788.552               2  
...        ...      ...             ...  
3995   8.66752  324.763               2  
3996  10.31806  380.140               2  
3997  10.32697  634.979               2  
3998   6.88789  621.063               2  
3999  11.77880  192.403               2  

[4000 rows x 10 columns]

In [43]:
#Kiểm tra dữ liệu
df1.shape

(4000, 10)

Dữ liệu có 4000 dòng và 10 cột

In [14]:
with open('./Coimbra_breast_cancer_dataset.txt', 'r') as file:
    description_dt = ""
    for i, line in enumerate(file):
        description_dt += f"{line.strip()}\n"
print(description_dt)

Quantitative Attributes:

Age (years): Represents the age of individuals in the dataset.

BMI (kg/m²): Body Mass Index, a measure of body fat based on weight and height.

Glucose (mg/dL): Reflects blood glucose levels, a vital metabolic indicator.

Insulin (µU/mL): Indicates insulin levels, a hormone associated with glucose regulation.

HOMA: Homeostatic Model Assessment, a method assessing insulin resistance and beta-cell function.

Leptin (ng/mL): Represents leptin levels, a hormone involved in appetite and energy balance regulation.

Adiponectin (µg/mL): Reflects adiponectin levels, a protein associated with metabolic regulation.

Resistin (ng/mL): Indicates resistin levels, a protein implicated in insulin resistance.

MCP-1 (pg/dL): Reflects Monocyte Chemoattractant Protein-1 levels, a cytokine involved in inflammation.

Labels:

1: Healthy controls

2: Patients with breast cancer

These quantitative attributes, including anthropometric data and parameters gathered from routine 

Thuộc tính định lượng:

1. Tuổi (năm): Thể hiện tuổi của các cá nhân trong tập dữ liệu.

2. BMI (kg/m2): Chỉ số khối cơ thể, thước đo lượng mỡ trong cơ thể dựa trên cân nặng và chiều cao.

3. Glucose (mg/dL): Phản ánh mức đường huyết, một chỉ số trao đổi chất quan trọng.

4. Insulin (µU/mL): Cho biết nồng độ insulin, một loại hormone liên quan đến điều hòa glucose.

5. HOMA: Đánh giá mô hình cân bằng nội môi, một phương pháp đánh giá tình trạng kháng insulin và chức năng tế bào beta.

6. Leptin (ng/mL): Thể hiện mức độ leptin, một loại hormone liên quan đến sự thèm ăn và điều chỉnh cân bằng năng lượng.

7. Adiponectin (µg/mL): Phản ánh nồng độ adiponectin, một loại protein liên quan đến điều hòa trao đổi chất.

8. resistin (ng/mL): Cho biết mức độ resistin, một loại protein có liên quan đến tình trạng kháng insulin.

9. MCP-1 (pg/dL): Phản ánh nồng độ Protein-1 của chất hóa học đơn nhân, một cytokine liên quan đến tình trạng viêm.

10. Nhãn: (1: Kiểm soát sức khỏe, 2: Bệnh nhân ung thư vú).



In [15]:
df.columns

Index(['Age', 'BMI', 'Glucose', 'Insulin', 'HOMA', 'Leptin', 'Adiponectin',
       'Resistin', 'MCP.1', 'Classification'],
      dtype='object')

In [16]:
df.info() #Thông tin DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             4000 non-null   int64  
 1   BMI             4000 non-null   float64
 2   Glucose         4000 non-null   int64  
 3   Insulin         4000 non-null   float64
 4   HOMA            4000 non-null   float64
 5   Leptin          4000 non-null   float64
 6   Adiponectin     4000 non-null   float64
 7   Resistin        4000 non-null   float64
 8   MCP.1           4000 non-null   float64
 9   Classification  4000 non-null   int64  
dtypes: float64(7), int64(3)
memory usage: 312.6 KB


In [17]:
df.describe() #Thống kê các cột

Age          BMI      Glucose      Insulin         HOMA  \
count  4000.00000  4000.000000  4000.000000  4000.000000  4000.000000   
mean     56.21075    27.422280   113.876500     8.654001     2.024332   
std      17.80965     4.413884    25.837795     6.435160     1.625638   
min      32.00000    20.690751    76.000000     2.821000     0.590033   
25%      39.00000    23.079053    76.000000     4.421750     0.970090   
50%      56.00000    27.558485   131.000000     5.818000     1.373842   
75%      72.00000    30.814916   134.000000    10.466250     2.502776   
max      85.00000    36.209606   138.000000    30.211000     8.218456   

            Leptin  Adiponectin     Resistin        MCP.1  Classification  
count  4000.000000  4000.000000  4000.000000  4000.000000     4000.000000  
mean     25.137737     9.364896    13.053667   512.183456        1.554000  
std      15.096446     4.674244     7.454424   253.279015        0.497138  
min       6.831900     3.192272     4.190320   137.488000        1.000000  
25%      12.712750     5.580210     7.022095   299.665000        1.000000  
50%      19.805050     8.286938    10.692780   482.308000        2.000000  
75%      36.670250    10.918292    17.294747   691.312500        2.000000  
max      68.506600    22.536919    42.617810  1227.250000        2.000000

In [18]:
df.isnull().sum() #Kiểm tra missing value

Age               0
BMI               0
Glucose           0
Insulin           0
HOMA              0
Leptin            0
Adiponectin       0
Resistin          0
MCP.1             0
Classification    0
dtype: int64

Dữ liệu không có giá trị NULL xuất hiện

In [19]:
df.duplicated().sum() #Kiểm tra mỗi hàng trong DataFrame có bị trùng lặp không

0

In [20]:
df[df.duplicated()]

Empty DataFrame
Columns: [Age, BMI, Glucose, Insulin, HOMA, Leptin, Adiponectin, Resistin, MCP.1, Classification]
Index: []

Dữ liệu không có bản ghi nào bị lặp lại hoàn toàn

In [9]:
data=pd.read_csv('Coimbra_breast_cancer_dataset.csv')
df1=data.copy()
df1

Age        BMI  Glucose  Insulin      HOMA   Leptin  Adiponectin  \
0      54  35.207389      103    5.642  1.378660  65.6699     9.738408   
1      52  22.978520      132    6.054  1.145435  47.5445     3.627241   
2      32  21.101341       87    5.668  1.008595  50.5074     5.067841   
3      42  26.761205      132    2.875  1.003837  16.8972    10.096475   
4      55  34.232520       76    3.120  1.597721  17.6852    11.845054   
...   ...        ...      ...      ...       ...      ...          ...   
3995   83  24.222881       76   10.511  0.733816  17.9836     5.423659   
3996   84  21.504832      137    4.567  1.314829  28.1327     5.878012   
3997   66  29.198300      137    6.823  0.688278   8.7539    10.242825   
3998   73  22.871102       90   10.564  3.623631  26.8907    14.407806   
3999   34  28.027362       76    3.008  1.235593  51.8521     7.984090   

      Resistin    MCP.1  Classification  
0     31.17499  197.660               1  
1     23.03327  423.361               2  
2      9.51156  897.034               2  
3      9.75652  267.744               1  
4     17.21541  788.552               2  
...        ...      ...             ...  
3995   8.66752  324.763               2  
3996  10.31806  380.140               2  
3997  10.32697  634.979               2  
3998   6.88789  621.063               2  
3999  11.77880  192.403               2  

[4000 rows x 10 columns]

In [10]:
X = df1.drop('Classification', axis=1)
y=np.array(df1['Classification'])

In [11]:
#Chuẩn hoá dữ liệu
standard_scaler = preprocessing.StandardScaler()
X_std = standard_scaler.fit_transform(X)

In [13]:
ratio = 0.20

In [14]:
#Chia tập dữ liệu 
X_train, X_test, y_train, y_test = train_test_split(X_std, y, test_size = ratio, shuffle = True, stratify = y)

In [36]:
print("Kích thước tập huấn luyện:", X_train.shape[0])
print("Kích thước tập kiểm tra:", X_test.shape[0])

Kích thước tập huấn luyện: 3200
Kích thước tập kiểm tra: 800


Tối ưu mô hình 
KNN

In [37]:
kf = KFold(n_splits=5)

In [38]:
train = kf.split(X_train, y_train)

In [39]:
fold_indices = []
for i, (train_index, test_index) in enumerate(train):
    fold_indices.append((train_index, test_index))

In [40]:
score_max = [0,0]
for k in range(3,100):
    model_kNN = KNeighborsClassifier(n_neighbors=k)
    score_k = cross_val_score(model_kNN,X_train,
               y_train, scoring = 'accuracy',
               cv = fold_indices).mean()
    if score_k > score_max[0]:
        score_max[0] = score_k
        score_max[1] = k

In [41]:
score_max

[0.534375, 89]

 Tại k = 89 thì mô hình thuật toán KNN có điểm cao nhất là 0.534375

In [4]:
model_kNN = KNeighborsClassifier(n_neighbors=89)

In [15]:
model_kNN.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=89)

In [48]:
y_pred=model_kNN.predict(X_test)

Đo độ đánh giá

In [51]:
# Tính toán precision, recall và F1-score
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
print(f'Precision:' , precision)
print(f'Recall: ',recall)
print(f'F1-score: ',f1)

Precision: 0.4631578947368421
Recall:  0.12324929971988796
F1-score:  0.1946902654867257
